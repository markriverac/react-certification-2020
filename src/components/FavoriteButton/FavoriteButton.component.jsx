import React, { useContext } from 'react';
import Button from 'react-bootstrap/Button';
import UserContext from '../../providers/UserProvider/User.context';
import useVideoStatus from '../../utils/hooks/useVideoStatus';

function FavoriteButton({ video }) {
  const { addFavorite, removeFavorite } = useContext(UserContext);
  const videoStatus = useVideoStatus(video);
  /* istanbul ignore next */
  function addNewFavorite() {
    addFavorite(video);
  }
  /* istanbul ignore next */
  function removeFromFavorites() {
    removeFavorite(video);
  }

  const buttonAdd = (
    <Button onClick={addNewFavorite} variant="outline-info">
      Add to favorite
    </Button>
  );
  const buttonRemove = (
    <Button onClick={removeFromFavorites} variant="outline-dark">
      Remove from Favorites
    </Button>
  );

  return videoStatus ? buttonRemove : buttonAdd;
}

export default FavoriteButton;
