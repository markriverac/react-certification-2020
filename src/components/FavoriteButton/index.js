/* istanbul ignore file */
export { default } from './FavoriteButton.component';
