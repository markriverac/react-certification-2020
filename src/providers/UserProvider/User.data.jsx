/* istanbul ignore file */
export const userData = { isAuth: false, name: '', favorites: [] };
