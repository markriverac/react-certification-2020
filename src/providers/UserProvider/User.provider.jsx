import React, { useState } from 'react';
import UserContext from './User.context';
import { userData } from './User.data';

function UserProvider({ children }) {
  const [userState, setuserState] = useState(userData);

  function logIn(username) {
    setuserState({ ...userState, name: username, isAuth: true });
  }
  function logOut() {
    setuserState({ ...userState, isAuth: false });
  }
  function addFavorite(video) {
    const newFavorites = userState.favorites;
    newFavorites.push(video);
    setuserState({ ...userState, favorites: newFavorites });
  }
  function removeFavorite(video) {
    const favoriteVideos = userState.favorites;
    const videoIndex = favoriteVideos.indexOf(video);
    if (videoIndex > -1) {
      favoriteVideos.splice(videoIndex, 1);
      setuserState({ ...userState, favorites: favoriteVideos });
    }
  }

  return (
    <UserContext.Provider
      value={{ userState, logIn, logOut, addFavorite, removeFavorite }}
    >
      {children}
    </UserContext.Provider>
  );
}

export default UserProvider;
