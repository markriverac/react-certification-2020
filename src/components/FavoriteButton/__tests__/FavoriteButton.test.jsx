import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import UserProvider from '../../../providers/UserProvider/User.provider';
import FavoriteButton from '../FavoriteButton.component';
import UserContext from '../../../providers/UserProvider/User.context';

Enzyme.configure({ adapter: new Adapter() });

it('Renders button without exploding', () => {
  const video = {};
  const button = renderer.create(
    <UserProvider>
      <FavoriteButton video={video} />
    </UserProvider>
  );
  const tree = button.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders add to favorite button', () => {
  const video = {};
  const button = renderer.create(
    <UserProvider>
      <FavoriteButton video={video} />
    </UserProvider>
  );
  const tree = button.toJSON();
  expect(tree.children).toEqual(['Add to favorite']);
});

it('renders remove from favorite button', () => {
  const video = {};
  const userState = { favorites: [video] };
  const button = renderer.create(
    <UserContext.Provider value={{ userState }}>
      <FavoriteButton video={video} />
    </UserContext.Provider>
  );
  const tree = button.toJSON();
  expect(tree.children).toEqual(['Remove from Favorites']);
});
