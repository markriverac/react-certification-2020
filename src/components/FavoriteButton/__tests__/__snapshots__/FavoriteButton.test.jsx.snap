// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders button without exploding 1`] = `
<button
  className="btn btn-outline-info"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Add to favorite
</button>
`;
